UPDATE Dockerfile to add bochs in container

Crescentm opened the pull request:
## 添加Bochs
build时构建bochs，bochs版本为2.8

bochs vnc图形化显示(vnc端口5900)，启动参数需加入端口映射 

`docker run -it -p 5900:5900 osfs_docker`

## 下面为最小demo

bochsrc
```
floppya: 1_44=hello.bin, status=inserted
boot: a
```

![CleanShot 2025-07-01 at 18 37 14@2x](https://github.com/user-attachments/assets/9a713961-933e-415c-9f98-3b54ffe2eb36)

Crescentm committed:
UPDATE Dockerfile to add bochs in container

Signed-off-by: Crescentm <guyues@outlook.com>

chenxiex commented on Dockerfile:
```diff
@@ -2,7 +2,11 @@ FROM ubuntu:16.04
+
+RUN cd /tmp &&  curl -L# https://sourceforge.net/projects/bochs/files/bochs/2.8/bochs-2.8.tar.gz | tar xz --strip-components=1
+RUN cd /tmp && ./configure --with-vncsrv --enable-debugger CXXFLAGS="-fpermissive" && make
+RUN mkdir -p /usr/local/share/bochs && cp -r /tmp/bios/* /usr/local/share/bochs && cp /tmp/bochs /usr/bin/bochs
```
直接使用`make install`是否更好？可以通过在`configure`时添加`--prefix=directory`更改安装位置。

chenxiex commented on Dockerfile:
```diff
@@ -12,4 +16,8 @@ RUN echo '#!/bin/bash\n/usr/bin/ld -m elf_i386 "$@"' > /usr/local/bin/ld && chmo
+RUN apt-get autoremove -y build-essential curl
+RUN rm -rf /var/lib/apt/lists/* && rm -rf /tmp/*
+EXPOSE 5900
+WORKDIR /mnt
```
设置`WORKDIR`似乎没有必要

chenxiex reviewed (changes requested):
当前似乎还不能直接运行 `osfs` 中的 `bochsrc`。主要有两点问题：
1. `keyboard_mapping` 选项在该版本的 bochs 中已经废弃。bochs 2.6 似乎支持这个选项。
2. `/usr/share/`下的一些文件找不到。

Crescentm commented:
> 当前似乎还不能直接运行 `osfs` 中的 `bochsrc`。主要有两点问题：
> 
> 1. `keyboard_mapping` 选项在该版本的 bochs 中已经废弃。bochs 2.6 似乎支持这个选项。
> 2. `/usr/share/`下的一些文件找不到。

1. 真的需要用这么老的bochsrc吗（武若叶的参考作业bochs版本是2.7）
2.已修改为make install的方式

Crescentm committed:
Update Dockerfile

- Downgrade bochs version from 2.8 to 2.6.11 for compatibility
- Use 'make install' for bochs instead of manual copy
- Remove WORKDIR /mnt

chenxiex commented on Dockerfile:
```diff
@@ -2,7 +2,10 @@ FROM ubuntu:16.04
 RUN dpkg --add-architecture i386 && \
     apt-get update && \
-    apt-get install -y gcc-multilib make nasm bear bc mtools git
+    apt-get install -y gcc-multilib make nasm bear bc mtools git libvncserver-dev libgcrypt-dev build-essential curl libgtk2.0-dev
```
这里添加 `apt-get install bochs`。编译的 bochs 默认会安装在`/usr/local/`，依然无法提供`/usr/share`下的文件。直接安装到`/usr`会容易和包管理的 bochs 冲突，如果用户后面又运行`apt install bochs`就炸了。干脆直接装两个 bochs，`/usr/local`下的bochs优先级更高会默认使用，包管理的 bochs 可以提供必要的文件和依赖，也可以通过`/usr/bin/bochs`手动调用（如果用户想要的话）。

Crescentm replied:
这样做会导致/usr/share下的依赖文件和bochs的版本不对应吧（以及这个不对应会造成多大影响？）

chenxiex replied:
确实会导致这个问题。目前不清楚会有什么影响，我测试过的都能跑。但我觉得这是相对比较好的解决方案了。另一个比较完美的解决方案是将编译的bochs打成deb包，依赖照抄bochs包，然后和原来的bochs包conflict。这样既可以安装到/usr，又可以避免文件脱离包管理导致的一系列问题。不过我对deb打包不熟，你有兴趣搞一下这个吗？

Crescentm replied:
我研究下怎么打包成deb，感觉这个方案确实优美

chenxiex commented on Dockerfile:
```diff
@@ -2,7 +2,10 @@ FROM ubuntu:16.04
-    apt-get install -y gcc-multilib make nasm bear bc mtools git
+    apt-get install -y gcc-multilib make nasm bear bc mtools git libvncserver-dev libgcrypt-dev build-essential curl libgtk2.0-dev
+
+RUN cd /tmp &&  curl -L# https://sourceforge.net/projects/bochs/files/bochs/2.6.11/bochs-2.6.11.tar.gz | tar xz --strip-components=1
```
我发现即使是 bochs 2.6.11 也无法使用 `keyboard_mapping` 选项。我在 osfs07上做了测试，发现直接注释掉这个选项也不影响运行，键盘也能正常输入。如果大家都喜欢用新版本的 bochs 的话，不如就切换回 2.8 的 bochs，然后写一个脚本，类似`patch_buildimg.sh`，把 `bochsrc` 里面的这一行直接注释掉。

Crescentm replied:
bochsrc 感觉可以写在README里（懒的写脚本了 😢 ）

Crescentm committed:
Dockerfile: upgrade bochs to 2.8 and install bochs via apt

- Install bochs via apt instead of building from source
- Update bochs tarball URL from 2.6.11 to 2.8

chenxiex commented:
因该 pr 已经可用，我先 merge 了。后续修改可提交新的 pr。
